xml/dhcp_vendor_class.xml

```diff
@@ -203,4 +213,93 @@
     <param pos="0" name="os.family" value="Windows"/>
   </fingerprint>
 
-</fingerprints>
\ No newline at end of file
+  <fingerprint pattern="^(?i)Cisco Systems Inc\. Wireless Phone (\d{4}g?)$" certainty="0.8">
+    <description>Cisco Wireless Phone</description>
+    <example hw.model="7920">Cisco Systems Inc. Wireless Phone 7920</example>
+    <example hw.model="7925G">Cisco Systems Inc. Wireless Phone 7925G</example>
+    <param pos="0" name="hw.device" value="VoIP"/>
+    <param pos="0" name="hw.vendor" value="Cisco"/>
+    <param pos="1" name="hw.model"/>
+    <param pos="0" name="hw.product" value="Unified Wireless IP Phone {hw.model}"/>
+    <param pos="0" name="os.vendor" value="Cisco"/>
+    <param pos="0" name="os.product" value="Unified Wireless IP Phone {hw.model} Firmware"/>
+  </fingerprint>
+
+  <fingerprint pattern="^Cisco systems, Inc\. IP Phone CP(39\d{2})$" certainty="0.8">
+    <description>Cisco SIP Phone</description>
+    <example hw.model="3911">Cisco systems, Inc. IP Phone CP3911</example>
+    <param pos="0" name="hw.device" value="VoIP"/>
+    <param pos="0" name="hw.vendor" value="Cisco"/>
+    <param pos="1" name="hw.model"/>
+    <param pos="0" name="hw.product" value="Unified SIP Phone {hw.model}"/>
+    <param pos="0" name="os.vendor" value="Cisco"/>
+    <param pos="0" name="os.product" value="Unified SIP Phone {hw.model} Firmware"/>
+  </fingerprint>
+
+  <fingerprint pattern="^Cisco Systems, Inc\. IP Phone CP-(\d{4}G?)(?:-GE)?$" certainty="0.8">
```

Comment:
I was taking a look at this fingerprint again after looking over the NVD CPEs and it appears the capture should include the value after the dash as a part of the model number.

For example, the value "Cisco Systems, Inc. IP Phone CP-7941G-GE" would result in the CPE `h:cisco:unified_ip_phone_7941g`, however, the model is "7941G-GE" and NVD contains the CPE `h:cisco:unified_ip_phone_7941g-ge`.

Reply:
As far as I can tell, the CPE contains the following entries for 7941:
```
h:cisco:unified_ip_phone_7941g
h:cisco:unified_ip_phone_7941g-ge

o:cisco:unified_ip_phone_firmware_7941g
o:cisco:unified_ip_phone_7941g_firmware
```

So both the G and GE models are present, but firmware only for G. I've no idea what the correct approach is (for instance, can anything bad happen if we refer to an inexistent firmware model?), so I'm totally happy to take guidance.

Reply:
@dmoinescu-r7 I understand the confusion. The CPEs from NVD are frequently added based on existing CVEs and sometimes reflect different naming decisions based on the individual that added them to the NVD database. In addition, understanding the reason for differences might require additional research into particular model(s) when possible. It appears we've identified an odd edge case with the 7941g model hardware and firmware. I would error on the side of emitting values that match the known model number. When we know we have the `7941g-ge` model we should emit values that match known CPEs with that model number. If we were to learn that the `7941g-ge` is simply a hardware revision and uses the `7941g` firmware we might carve off the `-ge` models into their own fingerprint allowing us more flexibility over formatting the params for those models. In general, one uses NVD data to guide formatting the various params. I would advise selecting the format that matches the greatest number in the NVD database. @TomSellers do you have anything you would like to add here?

Looking closer at the full list of `cisco:unified_ip_phone_` CPE it appears the fingerprint pattern could be extended to match two models that include `nr` after the numeric model.

```
h:cisco:unified_ip_phone_8851nr
h:cisco:unified_ip_phone_8865nr

o:cisco:unified_ip_phone_8851nr_firmware
o:cisco:unified_ip_phone_8865nr_firmware
```

From the perspective of new fingerprints there would be no failures caused by a reference to a nonexistent firmware model. It is simply data enrichment, but it is best to figure this out when the fingerprint is first added. As you discovered earlier the `update_cpes.py` script doesn't currently support strings that embed `{hw.model}`, therefore, it won't emit the `*.cpe23` params. Formatting the product params correctly now, will allow them to be emitted in the future when the tool is enhanced.

We should schedule a call this week to discuss the joys of CPE in greater detail. :)

Reply:
updated to allow for trailing NR